### ISSUE TITLE
[css-values-4] Should `<percentage>` be invalid in calculations matching `<number>`?

### ISSUE THREAD

**Issue by cdoublev** [w3c/csswg-drafts#9661]
I suspect that [determining the type](https://drafts.csswg.org/css-values-4/#determine-the-type-of-a-calculation) of `<percentage>` should be a failure when the grammar does not accept `<percentage>`, instead of defaulting to `[percent: 1]`:

  - `calc(1% / 1%)` should be an valid `z-index` because its grammar only accept `<number>`
  - `calc(1% / 1% * 1px)` should be a valid `width` because its grammar accepts both `<length>` and `<percentage>`, which resolves to a `<length>` based on the width or height of the containing block

Note: I assume that the type of `calc(1% / 1%)` as a value for `z-index` is currently `[percent: 0]`, which [matches](https://drafts.css-houdini.org/css-typed-om-1/#cssnumericvalue-match) `<number>`.

Or am I missing something?

**Comment by tabatkins** [comment]
No, the current behavior is somewhat intentional; it's similar to how `z-index: calc(2em / 10px)` is valid despite z-index not accepting lengths.

> Note: I assume that the type of calc(1% / 1%) as a value for z-index is currently [percent: 0], which [matches](https://drafts.css-houdini.org/css-typed-om-1/#cssnumericvalue-match) <number>.

Correct.

**Comment by cdoublev** [comment]
Relative lengths can always be resolved with an absolute base in `px`: `font-size` of the element, the viewport width, etc.

So `<percentage>`s in `1% / 1%` do not have any absolute base and are not resolved before applying the operation, which results to `1`, ok. Thanks.

**Comment by cdoublev** [comment]
Please note that step 9.4 of [simplify a calculation tree](https://drafts.csswg.org/css-values-4/#simplify-a-calculation-tree) requires to resolve `2em / 10px` with a value *expressed in the result’s canonical unit*, whereas this cannot be achieved (at parse time) with `2em`. I reported this problem in #7456 but at the time I created it, I was confused by other questions and preferred to close it.

Related to this issue and this problem: should `calc(1% / 1% * 1px)` (or `calc(1em / 1em * 1px)`) be resolved to `calc(1px)`? I would say no because when `1% === 0px`, the result is `NaN * 1px`. It all depends on whether relative numerics must be assumed to be resolved before resolving operations (except sums, eg. `2em - 1em` can be resolved to `1em`) or math functions (eg. `sign(1%)`).

Basically, I think products and math functions can be resolved (simplified) when they only include `<percentage>`s, if they do not resolve to another type (except `<number>`). They can not when they include relative dimensions.

**Comment by Loirooriol** [comment]
Related: #9667
